samd/adc_dac: Implememt adc.read_timed() and dac.write_timed().

robert-hh opened the pull request:
@dpgeorge As requested, this is the separate PR for ADC-timed and DAC-timed. Below
is a short documentation.

ADC Constructor:

ADC(Pin, *, average=16, bits=12, vref=3, callback=None)

The callback keyword option is used for timed ADC sampling. The callback is executed when all data has been sampled.

ADC Methods:

adc.read_u16()

Read single value from the Analog Pin using the bits and average setting from the constructors.

adc.read_timed(data, freq)

Read adc values into the data buffer at a supplied frequency. The buffer must be pre-allocated. Values are stored as 16 bit quantities in the binary range given by the bits option. If bits=12, the value range is 0-4095.
The voltage range is defined by the vref option.
If in the constructor a callback was defined, it will be called after all data has been read. Alternatively, the method busy() can be used to tell, if the capture has finished.

adc.busy()

busy() returns `True` while the data acquisition using read_timed() is ongoing, `False` otherwise. 

adc.deinit()

Deinitialize as ADC object and release the resources used by it, especially the ADC channel and the timer used for read_timed().


DAC Constructor

DAC(id, *, vref=3, callback=None)

The resolution of the DAC is 12 bit for SAMD51 and 10 bit for SAMD21. SAMD21 devices have 1 DAC channel at GPIO PA02, accepting only 0 as id. SAMD51 devices have 2 DAC channels at GPIO PA02 and PA05 with values 0 and 1 for the id. 

DAC Methods

dac.write(value)

Write a single value to the selected DAC output. The value range is 0-1023 for SAMD21 and 0-4095 for SAMD51. The voltage range depends on the vref setting.

dac.write_timed(data, freq [, count=1])

The call to dac_timed() allows to output a series of analogue values at a given rate. data must be a buffer with 16 bit values in the range of the DAC (10 bit of 12 bit). freq may have a range of 1Hz to ~200kHz for SAMD21 and 1 Hz to ~500kHz for SAMD51. The optional argument count specifies,
how often data output will be repeated. The range is 1 - 2**32. If count == 0, the data output will be repeated until stopped by a call to deinit(). If the data has been output count times, a callback will
be called, if given.

dac.busy()

Tells, whether a dac.write_timed() activity is ongoing. It returns `True` if yes, `False` otherwise.

dac.deinit()

Deinitialize the DAC and release the resources used by it, especially the DMA channel and the Timer. On most SAMD21 boards, there is just one timer available for dac.write_timed() and  adc.read_timed_into(). So they cannot run both at the same time, and releasing the timer may be important. The DAC driver consumes a substantial amount of current. Calling deinit()
will reduce that as well.

robert-hh committed:
samd/dma_manager: Add a DMA manager.

Used for allocation of DMA channels. It will be needed for planned
modules and methods like adc_timed(), dac_timed(), I2S.

It includes management code for DMA IRQ handlers, similar to what
was made for Sercom.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/tc_manager: Add a tc_manager function set.

These functions are use to allocate, free and configure a set
of TC counter instances. The SAMxx MCU have between 3 to 5 (SAMD21) and
4 to 8 (SAMD51) TC instances. Two of them are used for the µs counter,
the remaining 1 - 6 instances are administered here for use by
various functions, like timed DMA transfers.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/machine_dac: Add dac.write_timed() method.

Used as:

    dac.write_timed(data, freq [, count])
    dac.deinit()

Working range for dac_timed():

    SAMD21: 1 Hz - 100 kHz (1 MHz clock, 10 bit)
    SAMD51: 1 Hz - ~500 kHz (8 MHz clock, 12 bit)

The buffer has to be a byte array or a halfword array,
and the data is sent once.

The default for count is 1. If set to a value > 0, the data will be
transmitted count times. If set to 0 or  < 0, the date will be
transmitted until deliberately stopped. The playback
can be stopped with dac.deinit().

dac.deinit() just releases the timer and DMA channel needed by
dac_timed(). The DAC object itself does not have to be released.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/machine_adc: Add adc.read_timed() method.

Used as:

    adc.read_timed(buffer, freq)

Buffer must be preallocated. The size determines the number of 16 bit
words to be read. The numeric range of the results is that of the raw
ADC. The call returns immediately, and the data transfer is done by DMA.
The caller must wait sufficiently long until the data is sampled
and can be noticed by a callback. No internal checks are made for
a too-high freq value.

Read speeds depends on Average and bit length setting:

    SAMD21: Max. 350kS/s (8 bit, Average 1)
    SAMD51: Max. 1 MS/s (8 bit, Average 1)

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/machine_dac: Add a callback keyword option to machine.DAC().

The callback is called when a dac_timed() sequence finishes. It will be
reset with callback=None or omitting the callback option in the
constructor.

Side change: Set the clock freq. to 48Mhz.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/machine_adc: Add a callback keyword option to machine.ADC().

Enabling a callback that will be called when a adc.read_timed_into() run
is finished. That's especially useful with slow sampling rates and/or
many samples, avoiding to guess the sampling time.
Raise an error is adc.read_u16() is called while a read_timed_into()
is active.

Other ADC changes:
- SAMD51: use ADC1 if both ADC1 and ADC0 are available at a Pin.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd: Add new adc.busy() and dac.busy() methods.

These return True, while a timed action is ongoing.

Side change:
Reorder some code in machine_dac.c and do not reset DAC twice.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
docs/samd: Document the extensions to DAC and ADC.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/machine_dac: Rework the DAC deinit() semantics.

Since the two channels of a SAMD51 are not completely independent,
dac.deinit() now clears both channels, and both channels have to
be re-instantiated after a deinit().

Side change:
- rearrange some code lines.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd: Make adc.read_timed() and dac.write_timed() configurable.

Both together require ~1.9k of flash space, including the DMA-manager
and the TC-manager. adc.read_timed() uses ~700 bytes, dac.write_timed()
~600 bytes.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
Just a conflict resolved after rebase.

robert-hh commented:
@dpgeorge You mentioned about this PR that you would have to consider a proper API for that ADC/DAC timed feature. I guess you did not have time and it got lost under all other requests you have. The actual PR's API is close to that of the STM32 port.

robert-hh committed:
samd: Fix init and deinit for adc_timed() and dac_timed().

Fixes:
- Leave no half-initialized device if init fails.
- Fix dac_deinit_channel(). Perform deinit only for channels that
  had been initilized.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
Re-based and Re-tested. Still fine.

robert-hh committed:
samd/machine_adc: Factor out machine.adc_timed() to extmod code.

After machine.ADC has been moved to extmod/machine_adc.c.
Adding adc.read_timed() and adc.busy() to extmod/machine_adc.c with
a corresponding flag to enable them.
ADC/DAC timed are by default enabled only at all SAMD51 devices and
at SAMD21 devices with an external flash for the file system.

Add class constants for the reference voltage source.
As far as possible the STM32 names are used, except where they should
match common board silkscreen labels.

Signed-off-by: robert-hh <robert@hammelrath.com>

graeme-winter commented:
Is there hope of getting this merged into the mainline? Would be a useful addition - I tried cherry-picking the change set on top of #15890 but (unsurprisingly) there are quite a number of conflicts

robert-hh commented:
@graeme-winter I keep this PR in sync with the master, but do not update it on the sever if there is no conflict. But I did just now. Since your PR #15890 changes machine_dac.c as well, there are a few minor conflicts. So if one of the PR is merged, the other has to be changed accordingly.

dpgeorge commented:
I'm trying to test this PR on an ADAFRUIT_ITSYBITSY_M0_EXPRESS which is a SAMD21G18A (single DAC, only 3 TCs).

I have A0 and A1 wired together.  Using `adc.read_timed()` and then writing various values to the DAC using `dac.write()` seems to work well.

But this MCU only has 3 TCs and 2 are apparently used by the us counter, so I can't run both `DAC.write_timed` and `ADC.read_timed` at the same time.  Looking through the code, I can't see why 2 TCs are used by the us counter, it seems like only TC4 is used by it.  Eg in the isr vector table for SAMD21, only TC4 has an entry.  So, why can't TC5 be used by the DAC/ADC?

Regarding the data type for `ADC.read_timed()`, it writes a 16-bit wide value for each sample.  Is it possible to configure that to be 8 or 16-bits, depending on the type of the passed-in buffer?  stm32 has this feature, but I understand it might not be possible on samd because it's using DMA to read in the values.

robert-hh commented:
Thanks for looking into this. The SAMD21 chips currently supported by MicroPython have only 3 TC counters, of which 2 are used for the 32bit µs timer. So no TC4 or TC5.
I used an external signal generator and an oscilloscope for testing. You might use the PWM with an RC element to generate an analog signal for testing instead of tying DAC and ADC. The latter should work with SAMD51, but I never tested it.
I will look if ADC.read_timed() can be configured for 8 bit transfers. The DMA should be able to do 8 bit transfers, but I do not know how the ADC behaves when reading just one byte.

dpgeorge commented:
> of which 2 are used for the 32bit µs timer

Can you please show me in the code where and how it uses 2 timers?

robert-hh commented:
It is init_us_counter() in samd_soc.c. The samd21g18a has 3 TC counters, named TC3, TC4, TC5. TC4 and TC5 are combined to get a 32 bit counter.

robert-hh commented:
I made a few tests:
- There is a `LEFTADJ` flag in the configuration, which causes the result to be left-adjusted in the ADC result register. That works only for bits=10 and bits=12 with average=1. For bits=8 or average > 1 this flag has not effect. DMA access requires to read the either all bits of the result register or at least the upper byte, which is 0 in case of bits=8. Still, the ADC could be configured for 10 bit and then only the upper 8 bits could be taken, but that slows down the conversion and looks bad,.
- The existing code even for read_u16() does not deal properly with averaging. I made a fix fro `read_u16()`, but have to extend and test that with `read_timed()`. Till then, using average=1 is good for testing.

robert-hh committed:
samd/machine_adc: Fix the configuration with averaging enabled.

When averaging is selected, the resolution is fixed to 12 bit.
The configuration has to be changed to cater for the result shifts.

Side change: Remove a duplicated code line in init().

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
I added a commit to fix the average behavior.

dpgeorge commented:
> It is init_us_counter() in samd_soc.c. The samd21g18a has 3 TC counters, named TC3, TC4, TC5. TC4 and TC5 are combined to get a 32 bit counter.

OK, thanks for pointing that out.  Looking at the datasheet it says TC3 is paired with TC4, and that seems to be how `init_us_counter()` is implemented, but the comment for that function is wrong (says TC4/TC5).

It seems a shame to use two out of three timers for a microsecond counter.  Since SysTick is enabled and running it already produces a millisecond counter, and can also use that to get a microsecond counter like stm32 does it.

But, things work fine a they are, so no real need to change them (and no one asked for it so far).

dpgeorge commented on ports/samd/tc_manager.c:
```diff
@@ -0,0 +1,185 @@
+// List of channel flags: true: channel used, false: channel available
+// Two Tc instances are used by the usec counter and cannot be assigned.
+#if defined(MCU_SAMD21)
+static bool instance_flag[TC_INST_NUM] = {false, true, true};
```
Suggest using a simple integer with bit fields to track timer usage, eg `static uint8_t tc_instance_used`.

Then there could be a define like `#define TC_SYSTEM_USED 0x06` which defines which TCs are used by the system.  That makes `tc_deinit()` very simple: `tc_instance_used = TC_SYSTEM_USED`.

robert-hh replied:
I tried that with the TC-Manager. It caused a small increase in code size (16 resp 24 byte) and no change in RAM usage.

dpgeorge replied:
Interesting!  I guess the compiler already optimises it.

robert-hh commented:
> Looking at the datasheet it says TC3 is paired with TC4, and that seems to be how init_us_counter() is implemented, but the comment for that function is wrong (says TC4/TC5).

Actually TC4 and TC5 are combined. The datasheet seems to be wrong. Otherwise, TC3 could not be used.

> Since SysTick is enabled and running it already produces a millisecond counter, and can also use that to get a microsecond counter like stm32 does it.

But with a 1000µs resolution.

dpgeorge commented:
> But with a 1000µs resolution.

No, stm32's `mp_hal_ticks_us()` has 1us resolution.  That's because `SysTick->VAL` counts at the CPU frequency.  So it could potentially have resolution much better than 1us.

dpgeorge commented:
I tested this again on ADAFRUIT_ITSYBITSY_M0_EXPRESS, using ADC and DAC separately.

I used `DAC.write_timed(buf, n, 0)` to generate a continuous sine wave, measured using a scope, and it worked really well!

Testing `ADC.read_timed()` also looks good.

I think this is very close to being merged now.  @robert-hh is there anything more you want to do?  Some of the commits need to be squashed.  Feel free to do that yourself, or I can do it during merge.

dpgeorge commented on docs/samd/quickref.rst:
```diff
@@ -335,6 +385,52 @@ DAC Methods
+
+Deinitialize the DAC and release all resources used by it, especially the DMA channels
+and the Timers. On most SAMD21 boards, there is just one timer available for
+dac.write_timed() and adc.read_timed_into(). So they cannot run both at the same time,
```
`read_timed_into` -> `read_timed`

dpgeorge commented on extmod/machine_adc.c:
```diff
@@ -140,6 +140,24 @@ static mp_obj_t machine_adc_read(mp_obj_t self_in) {
 #endif
 
+#if MICROPY_PY_MACHINE_ADC_READ_TIMED
+// ADC.atten(value) -- this is a legacy method.
```
suggest: ADC.read_timed(buf, freq)

dpgeorge commented on extmod/machine_adc.c:
```diff
@@ -140,6 +140,24 @@ static mp_obj_t machine_adc_read(mp_obj_t self_in) {
+}
+static MP_DEFINE_CONST_FUN_OBJ_3(machine_adc_read_timed_obj, machine_adc_read_timed);
+
+// ADC.busy())
```
remove extra `)`

dpgeorge commented on ports/samd/mcu/samd21/mpconfigmcu.h:
```diff
@@ -76,6 +76,13 @@ unsigned long trng_random_u32(int delay);
+#define MICROPY_PY_MACHINE_ADC_READ_TIMED (SAMD21_EXTRA_FEATURES)
+#endif
+#ifndef MICROPY_PY_MACHINE_DAC_TIMED
+#define MICROPY_PY_MACHINE_DAC_TIMED    (SAMD21_EXTRA_FEATURES)
```
suggest calling this option `MICROPY_PY_MACHINE_DAC_WRITE_TIMED`

robert-hh commented:
@dpgeorge Thank you for the review. I incorporated the suggested changes. Reducing the number of commits seems difficult. I dropped two, and some could be combined, like two that add the callback option.

dpgeorge commented:
> Reducing the number of commits seems difficult.

OK, we can leave them mostly as they are.

The main thing I saw was the change to `extmod/machine_adc.c` was combined with other changes.  Ideally the additions to `extmod/machine_adc.c` would be in their own commit.

robert-hh commented:
> The main thing I saw was the change to extmod/machine_adc.c was combined with other changes. 

That is commit `1219b6ee6 2023-10-24 samd/machine_adc: Factor out machine.adc_timed().`, which by it's nature affects more than one file.

dpgeorge reviewed (approved):
OK, this is good to go!

dpgeorge commented:
Thanks @robert-hh for keeping this PR alive for so long, and updating it to latest master.

robert-hh commented:
Thank you very much for merging.
